lib/Sema/CSSimplify.cpp

```diff
@@ -3432,8 +3433,10 @@ performMemberLookup(ConstraintKind constraintKind, DeclName memberName,
     // See if we have an instance method, instance member or static method,
     // and check if it can be accessed on our base type.
     if (decl->isInstanceMember()) {
-      if ((isa<FuncDecl>(decl) && !hasInstanceMethods) ||
-          (!isa<FuncDecl>(decl) && !hasInstanceMembers)) {
+      if (((isa<FuncDecl>(decl) && !hasInstanceMethods) ||
+           (!isa<FuncDecl>(decl) && !hasInstanceMembers)) ||
+          (isMetatypeBase && !(functionRefKind == FunctionRefKind::Unapplied ||
+                               functionRefKind == FunctionRefKind::Compound))) {
```

Comment:
How about we replace this whole `if` statement with this:

```c++
      // If this is a method and either base is not allowed to have any,
      // or it's not going to be used as a value e.g. `let _ = X.foo`, report.
      if (isa<FuncDecl>(decl)) {
        auto isUnappliedOrCompoundRef = [](FunctionRefKind FRK) {
          return FRK == FunctionRefKind::Unapplied ||
                 FRK == FunctionRefKind::Compound;
        };

        if (!hasInstanceMethods ||
            (isMetatypeBase && !isUnappliedOrCompoundRef())) {
          result.addUnviable(candidate,
                             MemberLookupResult::UR_InstanceMemberOnType);
          return;
        }
      // If this is a property and base type is not allowed to have any, report.
      } else if (!hasInstanceMembers) {
        result.addUnviable(candidate,
                           MemberLookupResult::UR_InstanceMemberOnType);
        return;
      }
```

Reply:
Let me try!

Reply:
Can you please push your latest code so I can pull and test it a bit tonight?

Reply:
Done. What I find strange is that it it's getting added as `UR_InstanceMemberOnType` but `AllowTypeOrInstanceMemberFailure::diagnoseAsError()` isn't getting called. Very interesting!

Reply:
One more problem is that new logic in`performMemberLookup` is not going to work, because Swift support partial application, maybe counter-intuitive but this is supposed to work:

```swift
struct S {
  func foo() {}
}

var s = S()
S.foo(s) // S.foo returns (S) -> () -> () which then gets partially applied on instance of S via `s`
```

Reply:
@theblixguy Here are some changes to your PR https://gist.github.com/xedin/b004412cdbbaabb278fe3b46a154c727

I need to think about the best way to diagnose these partial application problems:

```swift
class A {}

enum B {
  static func foo() {
    bar(A())
  }

  func bar(_: A) {}
}
```

Currently our constraint system is setup poorly for this kinds of things...

Reply:
@theblixguy I think the only way forward we have is it diagnose these partial application errors via fixes before we can merge this (it's good anyway because at least one related to mutable base is diagnosed by MiscDiagnostics at the moment), I'll look into doing that in a bit.

Reply:
@xedin Does that mean we need another `FailureDiagnostic` subclass, etc?

Reply:
@xedin Oh, perhaps it can be done inside ```if (attemptFixes && kind >= ConstraintKind::Conversion) { ... }```, but the meta type is already stripped off. Or maybe in ```case ConstraintKind::Equal: { ... }``` we can check if `type2` is a `MetatypeType` and we can apply a fix & return `getTypeMatchSuccess()`? If I do that (w/o the fix) then the code throws `cannot invoke 'g' with an argument list of type '(Bar)'`, but I don't think such a simple check is correct.

Reply:
Ah never mind, just saw your PR!

Reply:
@xedin Okay 😄

Reply:
@xedin now that the partial applications PR has been merged, should I resume working on this? or are there any other upcoming changes that you're working on that I need to wait for?

Reply:
@xedin Oh! Are you working on another PR to diagnose those situations? or should that be handled in this PR?

Reply:
@xedin Oh, okay! I have synced the PR with master, so we can run a test now!